x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -93,12 +107,87 @@ protected void readInternal(byte[] b, int offset, int length) throws IOException
-            int read = inputStream.read(b, offset, length);
-            assert read == length;
-            position += read;
+    private void readInternalBytes(final int part, long pos, final byte[] b, int offset, int length) throws IOException {
+        int optimizedReadSize = readOptimized(part, pos, b, offset, length);
+        assert optimizedReadSize <= length;
+        position += optimizedReadSize;
+
+        if (optimizedReadSize < length) {
+            // we did not read everything in an optimized fashion, so read the remainder directly
+            try (InputStream inputStream
+                     = blobContainer.readBlob(fileInfo.partName(part), pos + optimizedReadSize, length - optimizedReadSize)) {
+                final int directReadSize = inputStream.read(b, offset + optimizedReadSize, length - optimizedReadSize);
+                assert optimizedReadSize + directReadSize == length : optimizedReadSize + " and " + directReadSize + " vs " + length;
+                position += directReadSize;
+            }
+        }
+    }
+
+    /**
+     * Attempt to satisfy this read in an optimized fashion using {@code streamForSequentialReadsRef}.
+     * @return the number of bytes read
+     */
+    private int readOptimized(int part, long pos, byte[] b, int offset, int length) throws IOException {
+        if (sequentialReadSize == NO_SEQUENTIAL_READ_OPTIMIZATION) {
+            return 0;
+        }
+
+        int read = 0;
+        if (streamForSequentialReads == null) {
+            // starting a new sequential read
+            read = readFromNewSequentialStream(part, pos, b, offset, length);
+        } else if (streamForSequentialReads.canContinueSequentialRead(part, pos)) {
+            // continuing a sequential read that we started previously
+            read = streamForSequentialReads.read(b, offset, length);
+            if (streamForSequentialReads.isFullyRead()) {
+                // the current stream was exhausted by this read, so it should be closed
+                streamForSequentialReads.close();
+                streamForSequentialReads = null;
+            } else {
+                // the current stream contained enough data for this read and more besides, so we leave it in place
+                assert read == length : length + " remaining";
+            }
+
+            if (read < length) {
+                // the current stream didn't contain enough data for this read, so we must read more
+                read += readFromNewSequentialStream(part, pos + read, b, offset + read, length - read);
+            }
+        } else {
+            // not a sequential read, so stop optimizing for this usage pattern and fall through to the unoptimized behaviour
+            assert streamForSequentialReads.isFullyRead() == false;
+            sequentialReadSize = NO_SEQUENTIAL_READ_OPTIMIZATION;
+            IOUtils.close(streamForSequentialReads);
+            streamForSequentialReads = null;
```

Comment:
We're closing + nullify the `streamForSequentialReads` many times, maybe it deserves its own `closeSequentialStream()` method?

Reply:
Good point, this is no longer a one-liner. Done in c9cf7bcd11d91a676f60d3a6bdb4c9b44858c368.